pkg/sources/s3/s3.go

```diff
@@ -204,6 +234,44 @@ func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) err
 		}
 	}
 	s.SetProgressComplete(len(bucketsToScan), len(bucketsToScan), fmt.Sprintf("Completed scanning source %s. %d objects scanned.", s.name, objectCount), "")
+	return nil
+}
+
+// Chunks emits chunks of bytes over a channel.
+func (s *Source) Chunks(ctx context.Context, chunksChan chan *sources.Chunk) error {
+	const defaultAWSRegion = "us-east-1"
+
+	if len(s.conn.Roles) > 0 {
+		for _, role := range s.conn.Roles {
+			client, err := s.newRoleClient(defaultAWSRegion, role)
+			if err != nil {
+				return errors.WrapPrefix(err, "could not create s3 client", 0)
+			}
+
+			bucketsToScan, err := s.getBucketsToScan(client)
+			if err != nil {
+				return err
+			}
+
+			if err := s.scanBuckets(ctx, client, role, bucketsToScan, chunksChan); err != nil {
+				return err
+			}
+		}
```

Comment:
I'm not very familiar with role based access, so forgive me if this is a dumb question.

Would it make sense to fallback to non-role based access if the roles were to fail?

Reply:
Also, this loop calls `scanBuckets` for each role? Does that mean we would scan the same data N times, or does each role have access to different data/buckets?

Reply:
The way that the SDK creates an authentication request makes fallback a bit cumbersome. The user essentially has to start the entire handshake process over again, so I don't think we'd gain much.

The functionality as-is will enumerate all the buckets a given role has the `s3:ListBucket` permission for and scan them, and then repeat for each role that has been passed in by the user.

Reply:
Ahh that makes a lot more sense! Thanks for clarifying for me.

Unless I'm still missing something, I think it would make sense to `continue` on error instead of returning so we can try to scan the next bucket. What do you think?